snuba/replacer.py

```diff
@@ -62,70 +62,50 @@ def get_connections(self) -> Mapping[int, Sequence[ClickhouseNode]]:
         raise NotImplementedError
 
 
-class RoundRobinConnectionPool(ShardedConnectionPool):
+class InOrderConnectionPool(ShardedConnectionPool):
     """
     Sharded connection pool that loads the valid nodes from the
     Clickhouse system.clusters table, then provides up to three
     connections for each shard picking them with a round robin
     policy.
 
-    TODO: Consider moving this logic and the executor to the
-          Clickhouse native module.
+    The list of connections provided is always the same so that
+    we send all writes to the first node of the shard and only
+    use the others as a fallback.
 
-    The goal is to evenly distribute the queries across the
-    nodes.
-
-    The sequence of connections for each shard is a sliding
-    window over the list of available connections. This window
-    wrap around when it reaches the end of the list and moves
-    by one node each time `get_connections` is invoked.
+    The rationale is to guarantee in order delivery. If all
+    queries are applied on the same node, they are guaranteed to
+    be applied in order. If they are applied on random nodes,
+    instead, the replication delay can make clickhouse apply
+    the updates out of order.
     """
 
     def __init__(
         self,
         cluster: ClickhouseCluster,
     ) -> None:
         self.__cluster = cluster
-        self.__counter = 0
         self.__nodes: Mapping[int, List[ClickhouseNode]] = defaultdict(list)
         self.__nodes_refreshed_at = time.time()
 
-    def __get_nodes(self) -> Mapping[int, Sequence[ClickhouseNode]]:
+    def get_connections(self) -> Mapping[int, Sequence[ClickhouseNode]]:
         now = time.time()
+
         if not self.__nodes or now - self.__nodes_refreshed_at > NODES_REFRESH_PERIOD:
             all_nodes = self.__cluster.get_local_nodes()
 
             self.__nodes = defaultdict(list)
-            # We pick up to three replicas per shard. The order will be the
-            # one provided by the get_local_nodes. For the correctness the order
-            # in which these nodes are tried does not matter.
-            for n in all_nodes:
-                if n.replica is not None and n.shard is not None:
+            # We pick up the first three replicas for each shard
+            sorted_nodes = sorted(
+                all_nodes, key=lambda n: n.replica if n.replica is not None else -1
+            )
+            for n in sorted_nodes:
+                if n.replica is not None and n.replica <= 3 and n.shard is not None:
```

Comment:
I don't get why n.replica <= 3 was added here. If it's here to limit the number of retries to 4 it should be commented and I think it would be better to do in the function that calls this and is responsible for the retry logic. This function's job is just to return all the valid nodes.

Reply:
I thought I remembered the contract of this class was to return 3 nodes at most, but I cannot find that constraint anymore so I must have remembered wrong. Will drop the constraint.